includes/Checker/Checks/Enqueued_Scripts_Size_Check.php

```diff
@@ -0,0 +1,229 @@
+	 *
+	 * @throws Exception Thrown when preparation fails.
+	 */
+	public function prepare() {
+		$orig_scripts = isset( $GLOBALS['wp_scripts'] ) ? $GLOBALS['wp_scripts'] : null;
+
+		// Backup the original values for the global state.
+		$check = $this;
+		$check->backup_globals();
+
+		return function() use ( $orig_scripts, $check ) {
+			if ( is_null( $orig_scripts ) ) {
+				unset( $GLOBALS['wp_scripts'] );
+			} else {
+				$GLOBALS['wp_scripts'] = $orig_scripts;
+			}
+
+			$check->restore_globals();
+		};
+	}
+
+	/**
+	 * Returns an array of shared preparations for the check.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
+	 *               need any constructor arguments, it would just be an empty array.
+	 */
+	public function get_shared_preparations() {
+		$demo_posts = array_map(
+			function( $post_type ) {
+				return array(
+					'post_title'   => "Demo {$post_type} post",
+					'post_content' => 'Test content',
+					'post_type'    => $post_type,
+					'post_status'  => 'publish',
+				);
+			},
+			$this->viewable_post_types
+		);
+
+		return array(
+			Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
+		);
+	}
+
+	/**
+	 * Runs the check on the plugin and amends results.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check results to amend and the plugin context.
+	 */
+	public function run( Check_Result $result ) {
+		$this->run_for_urls(
+			$this->get_urls(),
+			function ( $url ) use ( $result ) {
+				$this->check_url( $result, $url );
+			}
+		);
+	}
+
+	/**
+	 * Gets the list of URLs to run this check for.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array List of URL strings (either full URLs or paths).
+	 */
+	protected function get_urls() {
+		$urls = array( home_url() );
+
+		foreach ( $this->viewable_post_types as $post_type ) {
+			$posts = get_posts(
+				array(
+					'posts_per_page' => 1,
+					'post_type'      => $post_type,
+					'post_status'    => 'publish',
+				)
+			);
+			if ( isset( $posts[0] ) ) {
+				$urls[] = get_permalink( $posts[0] );
+			}
+		}
+
+		return $urls;
+	}
+
+	/**
+	 * Amends the given result by running the check for the given URL.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 * @param string       $url    URL to run the check for.
+	 *
+	 * @throws Exception Thrown when the check fails with a critical error (unrelated to any errors detected as part of
+	 *                   the check).
+	 */
+	protected function check_url( Check_Result $result, $url ) {
+		// Reset the WP_Scripts instance.
+		unset( $GLOBALS['wp_scripts'] );
+
+		// Run the 'wp_enqueue_script' action, wrapped in an output buffer in case of any callbacks printing scripts
+		// directly. This is discouraged, but some plugins or themes are still doing it.
+		ob_start();
+		wp_enqueue_scripts();
+		wp_head();
+		wp_footer();
```

Comment:
My main question about this approach is that it can introduce other side effects when all of this is called. I'm also unsure how this accounts for situations where someone is conditionally loading scripts based on some state associated with the URL, like the template type, etc.

Instead of calling wp_head() and wp_footer() directly, you could call `WP_Scripts::do_head_items()` and `WP_Scripts::do_footer_items()`, respectively without introducing other side effects if there are other actions hooked to header/footer actions.

Reply:
Thanks @joemcgill I agree it would be cleaner using the `WP_Scripts` methods you've mentioned. I've updated the PR to use these.

Reply:
Thanks @jjgrainger. Can you clarify my other question about how the `$url` is intended to be used in relationship to what scripts are being enqueued?

Reply:
@joemcgill we simulate the URL being requested by modifying the main `WP_Query`. We have a `URL_Aware` trait which handles this and provides methods to use within checks. The `run` method in this check uses those to loop over each of the URLs and run the check logic against each publicly viewable post type and home page. Let me know if you have any other questions.